mars/services/task/analyzer/fusion.py

```diff
@@ -12,101 +12,165 @@
-
-        return nodes_list, fused_nodes
-
-    def fuse(self) -> Tuple[List[List[ChunkType]], List[ChunkType]]:
-        fuses = []
-        explored = set()
-        # for those chunk in result sets, we should not do any fuse
-        result_chunk_set = set(self._graph.result_chunks)
-
-        for v in self._graph.topological_iter():
-            if v in explored or v in result_chunk_set:
-                continue
-            successors = self._graph.successors(v)
-            if len(successors) != 1:
-                continue
+    def __init__(
+        self,
+        chunk_graph: ChunkGraph,
+        all_bands: List[BandType],
+        chunk_to_bands: Dict[ChunkType, BandType],
+        initial_same_color_num: int = None,
+        as_broadcaster_successor_num: int = None,
+    ):
+        self.chunk_graph = chunk_graph
+        self.all_bands = all_bands
+        self.chunk_to_bands = chunk_to_bands
+        if initial_same_color_num is None:
+            initial_same_color_num = max(options.combine_size // 2, 1)
+        self.initial_same_color_num = initial_same_color_num
+        if as_broadcaster_successor_num is None:
+            as_broadcaster_successor_num = options.combine_size * 2
+        self.successor_same_color_num = as_broadcaster_successor_num
 
-            cur_node = successors[0]
-            if len(v.op.outputs) != 1:  # pragma: no cover
-                continue
-            if v.op.gpu != cur_node.op.gpu:
-                continue
-            if isinstance(v.op, (VirtualOperand, Fetch)):  # pragma: no cover
-                # cannot fuse virtual operand or fetch
-                continue
-            if (
-                v.op.scheduling_hint is not None
-                and not v.op.scheduling_hint.can_be_fused()
-            ):  # pragma: no cover
-                # don't fuse operand that cannot be fused
+        self._coloring_iter = itertools.count()
+
+    def _next_color(self) -> int:
+        return next(self._coloring_iter)
+
+    @classmethod
+    def _can_color_same(cls, chunk: ChunkType, predecessors: List[ChunkType]) -> bool:
+        if (
+            # VirtualOperand cannot be fused
+            any(isinstance(n.op, VirtualOperand) for n in [chunk] + predecessors)
+            # allocated on different bands
+            or len({n.op.gpu for n in [chunk] + predecessors}) > 1
+            # expect worker changed
+            or len({n.op.expect_worker for n in [chunk] + predecessors}) > 1
+            # scheduling hint tells that cannot be fused
+            or (
+                chunk.op.scheduling_hint is not None
+                and not chunk.op.scheduling_hint.can_be_fused()
+            )
+        ):
+            return False
+        return True
+
+    def _color_init_nodes(self) -> Dict[OperandType, int]:
+        # for initial op with same band but different priority
+        # we color them w/ different colors,
+        # to prevent from wrong fusion.
+        band_priority_to_colors = dict()
```

Comment:
A little strange using priority to do grouping here. Better to use distance measuring over undirected version of DAG to group initials.